File: Bulwark.Core/Engine/InjectionAnalyzer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Bulwark.Core.Models;

namespace Bulwark.Core.Engine;

/// <summary>
/// 进程注入 / DLL 侧载分析器(独创·内存与无文件攻击检测)。
///
/// 补足现有检测的一块空白:进程镂空(Process Hollowing)、APC 注入、线程劫持、
/// shellcode 注入、反射式 DLL 加载、DLL 搜索顺序劫持(侧载)等「无文件 / 内存型」手法。
/// 这些在内核/用户态事件里主要表现为两类信号:
///   · <see cref="EventType.RemoteThread"/>:一个进程向【另一个】进程创建远程线程
///     (镂空/APC/劫持/shellcode 注入的共同落点);
///   · <see cref="EventType.ImageLoad"/>:从可写/可疑目录加载未签名模块(侧载/搜索顺序劫持)。
///
/// 设计原则(遵循全局低误报):
///   · 注入【敏感高价值进程】(lsass / 系统关键进程 / 浏览器邮件)= 几乎无合法理由,置硬指标;
///   · 一般跨进程注入若发起方未签名 / 是脚本宿主或 LOLBin,也置硬指标;
///   · 带可信签名的合法程序做跨进程注入(反作弊 / IM / 录屏 / 安全软件本就会注入)仅给软信号,
///     交互证机制升格,避免误伤;
///   · 自注入(目标==自身)是常见合法行为,不计分;
///   · DLL 侧载仅对「高危可写目录(Temp/Public/ProgramData 根/回收站/PerfLogs)+ 未签名模块」
///     这一明确组合置硬指标,避免对从安装目录/AppData 加载自带 DLL 的正常程序误报。
///
/// 纯静态、无状态;返回 (score, reasons, hard) 供 <see cref="ThreatDetector"/> 汇总。
/// </summary>
public static class InjectionAnalyzer
{
    /// <summary>凭据宝库进程:向其注入几乎必为凭据窃取(T1003.001)。</summary>
    private static readonly HashSet<string> CredentialTargets =
        new(StringComparer.OrdinalIgnoreCase) { "lsass.exe", "lsaiso.exe" };

    /// <summary>关键系统进程:无合法理由被第三方注入。</summary>
    private static readonly HashSet<string> CriticalTargets =
        new(StringComparer.OrdinalIgnoreCase)
        {
            "winlogon.exe", "csrss.exe", "services.exe", "wininit.exe", "smss.exe",
            "svchost.exe", "spoolsv.exe", "explorer.exe", "dwm.exe", "lsm.exe",
        };

    /// <summary>浏览器/邮件进程:注入这些常为信息窃取 / 银行木马 / 会话劫持。</summary>
    private static readonly HashSet<string> SensitiveAppTargets =
        new(StringComparer.OrdinalIgnoreCase)
        {
            "chrome.exe", "msedge.exe", "firefox.exe", "iexplore.exe", "opera.exe",
            "brave.exe", "outlook.exe", "thunderbird.exe", "360se.exe", "360chrome.exe",
        };

    /// <summary>脚本宿主 / 常被滥用的 LOLBin:由它们发起注入是强信号。</summary>
    private static readonly HashSet<string> InjectorLolbins =
        new(StringComparer.OrdinalIgnoreCase)
        {
            "powershell.exe", "pwsh.exe", "cmd.exe", "wscript.exe", "cscript.exe",
            "mshta.exe", "rundll32.exe", "regsvr32.exe", "certutil.exe", "msbuild.exe",
        };

    /// <summary>高危可写落地目录:几乎只有恶意软件会在此释放并从中加载模块。</summary>
    private static readonly string[] HighRiskDirs =
    {
        @"\appdata\local\temp\", @"\windows\temp\", @"\users\public\",
        @"\programdata\", @"\$recycle.bin\", @"\perflogs\",
    };

    /// <summary>
    /// 分析一个事件的注入/侧载特征。仅对 RemoteThread / ImageLoad 有意义,其它返回零分。
    /// </summary>
    public static (int Score, List<string> Reasons, bool Hard) Analyze(SecurityEvent e)
    {
        if (e is null) return (0, new List<string>(), false);
        return e.Type switch
        {
            EventType.RemoteThread => AnalyzeRemoteThread(e),
            EventType.ImageLoad => AnalyzeImageLoad(e),
            _ => (0, new List<string>(), false)
        };
    }

    private static (int, List<string>, bool) AnalyzeRemoteThread(SecurityEvent e)
    {
        var reasons = new List<string>();
        if (string.IsNullOrEmpty(e.Target)) return (0, reasons, false);

        // 自注入(目标即自身)是常见合法行为(自身多进程/热补丁),不计分。
        if (string.Equals(e.Target, e.ActorPath, StringComparison.OrdinalIgnoreCase))
            return (0, reasons, false);

        string victim = SafeName(e.Target);
        string injector = SafeName(e.ActorPath);
        string injectorPathLower = (e.ActorPath ?? string.Empty).ToLowerInvariant();

        int score = 30; // 跨进程远程线程注入本身即值得关注
        bool hard = false;

        if (CredentialTargets.Contains(victim))
        {
            score += 50;
            hard = true;
            reasons.Add($"向凭据进程 {victim} 注入远程线程(疑似凭据窃取,T1003.001 / T1055)");
        }
        else if (CriticalTargets.Contains(victim))
        {
            score += 40;
            hard = true;
            reasons.Add($"向关键系统进程 {victim} 注入远程线程(无合法理由,T1055)");
        }
        else if (SensitiveAppTargets.Contains(victim))
        {
            score += 30;
            reasons.Add($"向浏览器/邮件进程 {victim} 注入远程线程(疑似信息窃取/会话劫持,T1055)");
        }
        else
        {
            reasons.Add($"跨进程远程线程注入 -> {victim}(进程镂空/APC/线程劫持的共同落点,T1055)");
        }

        bool injectorUnsigned = !e.ActorSigned;
        bool injectorIsLolbin = InjectorLolbins.Contains(injector);
        bool injectorInHighRiskDir = HighRiskDirs.Any(d => injectorPathLower.Contains(d));

        if (injectorIsLolbin)
        {
            score += 20;
            hard = true;
            reasons.Add($"注入发起方为脚本宿主/LOLBin({injector}),远程注入几乎必为恶意");
        }
        if (injectorUnsigned)
        {
            score += 15;
            // 未签名进程向另一个进程注入:正常软件几乎不做 -> 升为硬指标。
            hard = true;
            reasons.Add("注入发起方无可信签名");
        }
        if (injectorInHighRiskDir)
        {
            score += 10;
            reasons.Add("注入发起方位于高危可写目录");
        }

        // 说明:带可信签名、且非 LOLBin 的程序做一般跨进程注入(反作弊/录屏/IM/安全软件等)
        // 不置硬指标(hard 维持 false),仅累加软信号分,交互证机制升格,避免误伤合法注入。
        return (Math.Min(score, 100), reasons, hard);
    }

    private static (int, List<string>, bool) AnalyzeImageLoad(SecurityEvent e)
    {
        var reasons = new List<string>();
        string module = e.Target ?? string.Empty;
        if (string.IsNullOrEmpty(module)) return (0, reasons, false);

        // 注意:ImageLoad 事件里 ActorSigned 表示【被加载模块】自身的签名(由事件源按模块判定)。
        bool moduleSigned = e.ActorSigned;
        string moduleLower = module.ToLowerInvariant();
        bool inHighRiskDir = HighRiskDirs.Any(d => moduleLower.Contains(d));

        // 明确组合:未签名模块 + 从高危可写目录加载 = DLL 侧载/搜索顺序劫持的典型形态。
        // 合法程序极少从 Temp/Public/ProgramData 根加载未签名 DLL。
        if (!moduleSigned && inHighRiskDir)
        {
            reasons.Add($"从高危可写目录加载未签名模块 {SafeName(module)}(疑似 DLL 侧载/搜索顺序劫持,T1574.002)");
            return (40, reasons, true);
        }

        return (0, reasons, false);
    }

    private static string SafeName(string? path)
    {
        if (string.IsNullOrEmpty(path)) return string.Empty;
        try { return Path.GetFileName(path).ToLowerInvariant(); }
        catch { return path.ToLowerInvariant(); }
    }
}

File: cpp/shared/src/engine/InjectionAnalyzer.cpp
#include "bulwark/engine/InjectionAnalyzer.h"
#include <QSet>
#include <QStringList>

namespace bulwark::engine {
using detail::u;
using detail::fileNameLower;

namespace {

const QSet<QString>& credentialTargets() {
    static const QSet<QString> s = { "lsass.exe", "lsaiso.exe" };
    return s;
}
const QSet<QString>& criticalTargets() {
    static const QSet<QString> s = {
        "winlogon.exe", "csrss.exe", "services.exe", "wininit.exe", "smss.exe",
        "svchost.exe", "spoolsv.exe", "explorer.exe", "dwm.exe", "lsm.exe",
    };
    return s;
}
const QSet<QString>& sensitiveAppTargets() {
    static const QSet<QString> s = {
        "chrome.exe", "msedge.exe", "firefox.exe", "iexplore.exe", "opera.exe",
        "brave.exe", "outlook.exe", "thunderbird.exe", "360se.exe", "360chrome.exe",
    };
    return s;
}
const QSet<QString>& injectorLolbins() {
    static const QSet<QString> s = {
        "powershell.exe", "pwsh.exe", "cmd.exe", "wscript.exe", "cscript.exe",
        "mshta.exe", "rundll32.exe", "regsvr32.exe", "certutil.exe", "msbuild.exe",
    };
    return s;
}
const QStringList& highRiskDirs() {
    static const QStringList s = {
        "\\appdata\\local\\temp\\", "\\windows\\temp\\", "\\users\\public\\",
        "\\programdata\\", "\\$recycle.bin\\", "\\perflogs\\",
    };
    return s;
}
bool anyContains(const QStringList& needles, const QString& hay) {
    for (const QString& n : needles)
        if (hay.contains(n)) return true;
    return false;
}

//
// 注入发起方是否为「Windows 自身的系统组件」:签名健康 + 位于系统目录 + 不是 LOLBin/脚本宿主。
//
// 补的是一处实测误报:本分析器原来只豁免【自注入】(target == actorPath),`criticalTargets()`
// 里一命中就 +40 且置硬指标,【完全不看谁在注入】。于是 Windows 内部的正常线程创建被当成攻击:
//   · C:\Windows\System32\dwm.exe     -> csrss.exe   85 分被拦(12 次)
//   · C:\Windows\System32\svchost.exe -> csrss.exe   86 分被拦
// 两者都是 Microsoft 签名、位于 System32。桌面窗口管理器和服务宿主与 csrss 之间有大量正常的
// 跨进程线程交互(窗口/会话管理),把它算成 T1055 注入是判据缺了「注入方身份」这个前置条件。
//
// 判据刻意收得很紧,避免变成绕过口子:
//   * 必须签名健康(actorSigned 且未失配 / 未吊销 / 未过期后签名)—— 改名成 dwm.exe 丢在
//     Temp 里的样本不满足;
//   * 必须位于 System32 / SysWOW64 / WinSxS —— 这些目录普通账户写不进去;
//   * 【排除 LOLBin 与脚本宿主】—— powershell / rundll32 / mshta 也签名、也在 System32,
//     但它们做远程注入没有任何正当理由,那正是本分析器最该抓的形态,绝不能连带豁免。
// 三条同时成立时,才认为「这是系统自己在做自己的事」,不再单独定罪(分数照常累加,仍可被
// 其它硬指标互证升格)。
//
bool injectorIsTrustedOsComponent(const bulwark::SecurityEvent& e) {
    if (!e.actorSigned) return false;
    if (e.signatureMismatch || e.certRevoked || e.signedAfterCertExpiry) return false;

    const QString injector = fileNameLower(e.actorPath);
    if (injectorLolbins().contains(injector)) return false;

    QString p = e.actorPath.toLower();
    p.replace(QLatin1Char('/'), QLatin1Char('\\'));
    return p.contains(QLatin1String("\\windows\\system32\\"))
        || p.contains(QLatin1String("\\windows\\syswow64\\"))
        || p.contains(QLatin1String("\\windows\\winsxs\\"));
}

ScoreResult analyzeRemoteThread(const bulwark::SecurityEvent& e) {
    ScoreResult r;
    if (e.target.isEmpty()) return r;
    // 自注入(目标==自身)是常见合法行为,不计分。
    if (e.target.compare(e.actorPath, Qt::CaseInsensitive) == 0) return r;

    const QString victim = fileNameLower(e.target);
    const QString injector = fileNameLower(e.actorPath);
    const QString injectorPathLower = e.actorPath.toLower();

    int score = 30; // 跨进程远程线程注入本身即值得关注
    bool hard = false;

    // 「系统组件之间的线程交互」与「攻击者注入系统进程」必须分开(见 injectorIsTrustedOsComponent)。
    const bool osComponent = injectorIsTrustedOsComponent(e);

    if (credentialTargets().contains(victim)) {
        // 凭据进程(lsass / lsaiso)【不】给系统组件豁免:代价太高,宁可让极少数正常交互走一次询问。
        score += 50; hard = true;
        r.reasons << (u("向凭据进程 ") + victim + u(" 注入远程线程(疑似凭据窃取,T1003.001 / T1055)"));
    } else if (criticalTargets().contains(victim)) {
        if (osComponent) {
            // 签名健康的系统目录组件(且非 LOLBin)与另一个系统进程之间的线程创建:
            // Windows 内部的窗口 / 会话 / 服务管理本来就这么工作(实测 dwm.exe、svchost.exe
            // 对 csrss.exe)。保留少量分数留痕、可被其它硬指标互证,但绝不单独定罪。
            score += 10;
            r.reasons << (u("系统组件 ") + injector + u(" 向系统进程 ") + victim +
                          u(" 创建线程(Windows 内部交互的常见形态,仅留痕)"));
        } else {
            score += 40; hard = true;
            r.reasons << (u("向关键系统进程 ") + victim + u(" 注入远程线程(无合法理由,T1055)"));
        }
    } else if (sensitiveAppTargets().contains(victim)) {
        score += 30;
        r.reasons << (u("向浏览器/邮件进程 ") + victim + u(" 注入远程线程(疑似信息窃取/会话劫持,T1055)"));
    } else {
        r.reasons << (u("跨进程远程线程注入 -> ") + victim + u("(进程镂空/APC/线程劫持的共同落点,T1055)"));
    }

    const bool injectorUnsigned = !e.actorSigned;
    const bool injectorIsLolbin = injectorLolbins().contains(injector);
    const bool injectorInHighRiskDir = anyContains(highRiskDirs(), injectorPathLower);

    if (injectorIsLolbin) {
        score += 20; hard = true;
        r.reasons << (u("注入发起方为脚本宿主/LOLBin(") + injector + u("),远程注入几乎必为恶意"));
    }
    if (injectorUnsigned) {
        score += 15; hard = true;
        r.reasons << u("注入发起方无可信签名");
    }
    if (injectorInHighRiskDir) {
        score += 10;
        r.reasons << u("注入发起方位于高危可写目录");
    }

    r.score = qMin(score, 100);
    r.hardSignal = hard;
    return r;
}

ScoreResult analyzeImageLoad(const bulwark::SecurityEvent& e) {
    ScoreResult r;
    const QString module = e.target;
    if (module.isEmpty()) return r;

    // ImageLoad 事件里 actorSigned 表示【被加载模块】自身的签名。
    const bool moduleSigned = e.actorSigned;
    const QString moduleLower = module.toLower();
    const bool inHighRiskDir = anyContains(highRiskDirs(), moduleLower);

    if (!moduleSigned && inHighRiskDir) {
        r.reasons << (u("从高危可写目录加载未签名模块 ") + fileNameLower(module) +
                      u("(疑似 DLL 侧载/搜索顺序劫持,T1574.002)"));
        r.score = 40;
        r.hardSignal = true;
    }
    return r;
}

} // namespace

ScoreResult InjectionAnalyzer::analyze(const bulwark::SecurityEvent& e) {
    switch (e.type) {
        case EventType::RemoteThread: return analyzeRemoteThread(e);
        case EventType::ImageLoad:    return analyzeImageLoad(e);
        default:                      return ScoreResult{};
    }
}

} // namespace bulwark::engine
